docs: add new migration guide to v12 documentation that describes how… (#41828)
angular/angular: aio/content/guide/migration-legacy-message-id.md

# Migrating legacy localization IDs

## What this migration does

> Angular version 12 introduces tooling to help you migrate any existing localization ids to ids that use the latest algorithms.

Reviewer: Since v11, all new projects use the modern IDs by default.
Also any use of `$localize` in application code also uses the moden ID format.
Further, some projects use custom IDs (the set the ID for each translation manually themselves). That being said any legacy ICU messages will not have custom IDs and will need to be migrated.

So, strictly, this migration might not be necessary. Or at least when describing it we should be clear that we only need to migrate "legacy" IDs rather than all existing ones.

Reviewer 2: Thanks for this clarification, @petebacondarwin !

> You have two options for migrating legacy IDs. The first method locates existing IDs, creates new IDs, and replaces the old IDs. The second method is for cases custom localization processes, such as where your localization IDs are not stored in the Angular application. Use this second method if you store your IDs in a database. This method locates legacy IDs and then creates a set of new IDs that you can then use to replace the old ones.

Reviewer: The only difference (really) between the two solutions is that the first (and by far the most common) will be to use the CLI to locate existing IDs. The other solution is to use the standalone binary `localize-extract` to locate the existing IDs.

Both use the same second step, which is to run `localize-migrate` to update translation files.

## Create and replace legacy IDs

To create new localization IDs and replace the old ones:

1. Run the `ng extract-i18n` command.

   `ng extract-i18n --format=legacy-migrate`

1. Next, run the `npx localize-migrate` command.

>    `npx localize-migrate --files=*.xlf --mapFile=messages.json`

Reviewer: We should be careful here. Although the vast majority of people use XLIFF translation files, they may have their translations stored in other formats such as XMB or (very rarely JSON).

Reviewer 2: I added a note to reflect this. Let me know if I'm missing the key idea.

1. Commit the updated files to your source control system.

## Migrate localization IDs for custom build processes

1. Run the `npx localize-extract` command.

   `npx localize-extract --format=legacy-migrate --source=./path/to/bundles --outputPath=./`

   In this command, `./path/to/bundles/` represents the path to your localization files. You can set the `outputPath` parameter to any directory in your system.

1. Next, run the `npx localize-migrate` command.

   `npx localize-migrate --files=*.xlf --mapFile=messages.json`

> You now have a list of legacy localization IDs and new IDs, which you can use to update your application.

Reviewer: This line should actually be part of the previous step. I.e. after you have run `localize-extract` you now have a list of legacy IDs (the mapfile) that can be used to update the IDs in your translation files.

After running `localize-migrate` your translation files should now have been updated to use the new IDs. There should be nothing more to do on the application side.

# Migrating legacy localization IDs

## What this migration does

> 

Reviewer 3: I think that we should mention somewhere that this is only a problem if people aren't using custom IDs. I know that there are some applications out there that enforce custom IDs everywhere and which won't need to migrate.

> Angular version 11 introduced a new format for localization IDs. These new IDs are more robust the previous, legacy format. However, applications created before version 11 still used the legacy format for these IDs.

Reviewer 4: ```suggestion
Angular version 11 introduced a new format for localization IDs. These new IDs are more robust than the previous, legacy format. However, applications created before version 11 still used the legacy format for these IDs.
```

Reviewer 2: Thank you, @ahasall !

> Angular version 11 introduced a new format for localization IDs. These new IDs are more robust the previous, legacy format. However, applications created before version 11 still used the legacy format for these IDs.

Reviewer 5: ```suggestion
Angular version 11 introduced a new format for localization IDs. These new IDs are more robust the previous legacy format. However, applications created before version 11 still used the legacy format for these IDs.
```

With the release of Angular version 12, you now how tools available to help you migrate any existing localization ids to ids that use the latest algorithms.

> You have two options for migrating legacy IDs. The first method uses the Angular CLI to locate legacy IDs in your application. The second method uses a standalone application, `localize-extract` to locate the legacy IDs.

Reviewer 5: ```suggestion
You have two options for migrating legacy IDs. The first method uses the Angular CLI to locate legacy IDs in your application. The second method uses a standalone application, `localize-extract`, to locate the legacy IDs.
```

## Migrating legacy IDs using the CLI

To migrate legacy localization IDs using the CLI:

1. Run the `ng extract-i18n` command.

   `ng extract-i18n --format=legacy-migrate`

   After running this command, you have a list of legacy localization IDs and new IDs, which you can use to update your application.

1. Update the IDs in your application using the `npx localize-migrate` command.

>    `npx localize-migrate --files=*.xlf --mapFile=messages.json`

Reviewer: There is a slight problem with using `npx`.
First we would need to specify the package (e.g. `-p @angular/localize`) as well as the command `localize-migrate`.
But also since `@angular/localize` only has `@angular/compiler` and `tslib` etc as peer dependencies, `npx` will not install them, so the whole thing fails.

Instead we should get them to run the local version of this executable (since they will have `@angular/localize` installed in their project already). This is easy with yarn: `yarn localize-migrate` but that is not possible with npm, in which case you would need to specify the path directly: `node_modules/.bin/localize-migrate`.

Reviewer 2: Can you give me the specific steps you want users to follow? I'm not sure I understand what they are.

Reviewer 4: @petebacondarwin, if `@angular/localize` is installed locally then `npx localize-migrate` should work fine and is equivalent to `./node_modules/.bin/localize-migrate` I think.

Reviewer: OK let’s go with the verbose option, which will work in all cases (I believe)

```
node_modules/.bin/localize-migrate --files=*.xlf --mapFile=messages.json
```

Of course if they “don’t” have @angular/localize installed in their project, this would fail too. But if they don’t have it installed then they can’t be doing translations, so :shrug:

Similarly later on for the non-CLI version we need:

```
node_modules/.bin/localize-extract --format=legacy-migrate --source=./path/to/bundles --outputPath=./messages.json
```

Reviewer: > @petebacondarwin, if `@angular/localize` is installed locally then `npx localize-migrate` should work fine and is equivalent to `./node_modules/.bin/localize-migrate` I think.

That's interesting... I thought I had tried this in a local project and it still failed. 

@crisbeto - do you have any thoughts on this?

Reviewer 3: I think that @ahasall is right that it should work if `@angular/localize` is installed already. As I was developing the script, I was testing against a simple CLI app and I was using `npx` for it.

Reviewer: In that case, you can ignore my comments about `npx` and move on.

Reviewer 2: I will leave this content unchanged, then. Let me know if I am misinterpreting anything!

   <div class="alert is-helpful">

   You can also specify other formats in the `files` parameter, such as `*.xmb`.
   </div>

1. Commit the updated files to your source control system.

## Migrate legacy IDs using `localize-extract`

> To migrate legacy localization IDs using `localize-extract`:

Reviewer: How about:

```suggestion
If you are not using the Angular CLI, then you can migrate legacy localization IDs using `localize-extract`:
```

1. Run the `npx localize-extract` command.

>    `npx localize-extract --format=legacy-migrate --source=./path/to/bundles --outputPath=./`

Reviewer: The `outputPath` has to actually be a path to the file, not the directory. So

```suggestion
   `npx localize-extract --format=legacy-migrate --source=./path/to/bundles --outputPath=./messages.json`
```

>    In this command, `./path/to/bundles/` represents the path to your localization files. You can set the `outputPath` parameter to any directory in your system.

Reviewer: ```suggestion
   In this command, `./path/to/bundles/` represents the path to your distributable files. You can set the `outputPath` parameter to any directory in your system.
```

The bundles will be the generated JavaScript from whatever build tooling they are using.

>    After running this command, you have a list of legacy localization IDs and new IDs, which you can use to update your application.

Reviewer: ```suggestion
   After running this command, you have a migration file (`messages.json`) containing a mapping between legacy localization IDs and new IDs, which you can use to update your application.
```

1. Update the IDs in your application using the `npx localize-migrate` command.

   `npx localize-migrate --files=*.xlf --mapFile=messages.json`

   <div class="alert is-helpful">

   You can also specify other formats in the `files` parameter, such as `*.xmb`.

>    </div>

Reviewer: ```suggestion

   </div>
```

> 

Reviewer: That's a good idea!
And of course, it is important that it is set to true when the mapping file is generated :-)

> 

Reviewer 2: Added some content about this. PTAL!

> ## Why is this migration necessary

Reviewer 5: ```suggestion
## Why this migration is necessary
```
Moved word for better flow

> The Angular translation system works by matching a message ID to a translated message in a different language. One option for creating these IDs is to have the translation system generate them for you. Previous algorithms that created these IDs have, over time, become fragile to formatting changes. The new algorithms are much more robust. This migration system helps you improve application stability by creating new legacy IDs that you can add to your application.

Reviewer 3: In the last sentence, I'm not sure that "application stability" is the correct term. Migrating to the new system makes it more future-proof, because they won't have to migrate later when we remove the new format. It also means that the IDs will change less frequently.

> The Angular translation system works by matching a message ID to a translated message in a different language. One option for creating these IDs is to have the translation system generate them for you. Previous algorithms that created these IDs have, over time, become fragile to formatting changes. The new algorithms are much more robust. This migration system helps you improve application stability by creating new legacy IDs that you can add to your application.

Reviewer 6: It seems like it would be useful to move the content from "why is this migration necessary" higher up so that people more quickly know why they should care about this migration in the first place.

Reviewer 2: I was following the pattern of previous migration guides, but I agree. I'll put this section ahead of "What this migration does"

# Migrating legacy localization IDs

## Why this migration is necessary

> The Angular translation system works by matching a message ID to a translated message in a different language. One option for creating these IDs is to have the translation system generate them for you. Previous algorithms that created these IDs have, over time, become fragile to formatting changes. The new algorithms are much more robust. This migration system helps you future-proof your application in the event that the old format is removed.

Reviewer 2: Made some changes that I think are in the spirit of what you're asking for. PTAL.

## What this migration does

Angular version 11 introduced a new format for localization IDs. These new IDs are more robust than the previous legacy format. However, applications created before version 11 still used the legacy format for these IDs.

<div class="alert is-helpful">

> If you use custom localization IDs in your applications, you do not need to migrate to new IDs.

Reviewer 2: Again, made a change that's in the spirit of this. PTAL.

</div>

> With the release of Angular version 12, you now have how tools available to help you migrate any existing localization IDs to IDs that use the latest algorithms.

Reviewer 5: ```suggestion
With the release of Angular version 12, you now have tools available to help you migrate any existing localization IDs to IDs that use the latest algorithms.
```
Deleted unnecessary word.

You have two options for migrating legacy IDs. The first method uses the Angular CLI to locate legacy IDs in your application. The second method uses a standalone application, `localize-extract`, to locate the legacy IDs.

## Migrating legacy IDs using the CLI

To migrate legacy localization IDs using the CLI:

1. Run the `ng extract-i18n` command.

   `ng extract-i18n --format=legacy-migrate`

   After running this command, you have a migration file, `messages.json`, containing a mapping between legacy localization IDs and new IDs that you can use to update your application.

1. Update the IDs in your application using the `npx localize-migrate` command.

   `npx localize-migrate --files=*.xlf --mapFile=messages.json`

   <div class="alert is-helpful">

   You can also specify other formats in the `files` parameter, such as `*.xmb`.

   </div>

1. Commit the updated files to your source control system.

> After you complete the migration, set the Angular Compiler option, `enableI18nLegacyMessageIdFormat`, to `false`. For more information about this option, see [Angular Compiler Options](/guide/angular-compiler-options#enablei18nlegacymessageidformat).

Reviewer 2: No, because these are two different sections. Folks who are reading one may not read the other.
